sdk/core/System.ClientModel/src/Message/PipelineMessage.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Threading;
+
+namespace System.ClientModel.Primitives;
+
+// TODO: MessageClassifier and RequestOptions will come in a later PR.
+
+public class PipelineMessage : IDisposable
+{
+    private ArrayBackedPropertyBag<ulong, object> _propertyBag;
+    private bool _disposed;
+
+    protected internal PipelineMessage(PipelineRequest request)
+    {
+        Argument.AssertNotNull(request, nameof(request));
+
+        Request = request;
+        _propertyBag = new ArrayBackedPropertyBag<ulong, object>();
+    }
+
+    public PipelineRequest Request { get; }
+
+    public PipelineResponse? Response
+    {
+        get;
+
+        // Set internally by the transport.
+        protected internal set;
+    }
+
+    #region Pipeline invocation options
+
+    public CancellationToken CancellationToken { get; set; }
+
+    public bool TryGetProperty(Type type, out object? value) =>
+        _propertyBag.TryGetValue((ulong)type.TypeHandle.Value, out value);
+
+    public void SetProperty(Type type, object value) =>
+        _propertyBag.Set((ulong)type.TypeHandle.Value, value);
+
+    #endregion
+
+    #region Meta-data for pipeline processing
+
+    internal int RetryCount { get; set; }
+
+    #endregion
+
+    #region Per-request pipeline
+
+    internal bool CustomRequestPipeline =>
```

Comment:
```suggestion
    internal bool IsCustomRequestPipeline =>
```

Justification: The member is used to determine if this is a custom request pipeline, not access the custom request pipeline.

Reply:
It's an internal API, so I'm less concerned about naming, but happy to make the change to clarify intent for maintainers!